src/kbmod/search/geom.h

```diff
@@ -0,0 +1,226 @@
+    }
+
+    const std::string to_yaml() const {
+        return "{x: " + std::to_string(y) + " y: " + std::to_string(x) + "}";
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const Point& rc);
+};
+
+std::ostream& operator<<(std::ostream& os, const Point& rc) {
+    os << rc.to_string();
+    return os;
+}
+
+// A rectangle that also contains it's corner's origin Index with respect to
+// another, larger rectangle.
+struct AnchoredRectangle {
+    Index corner;
+    Index anchor;
+
+    unsigned height;
+    unsigned width;
+
+    const std::string to_string() const {
+        return "AnchoredRectangle(corner:" + corner.to_string() + ", anchor:" + anchor.to_string() +
+               ", height: " + std::to_string(height) + ", width: " + std::to_string(width) + ")";
+    }
+
+    const std::string to_yaml() const {
+        return "{corner: " + corner.to_yaml() + " anchor: " + anchor.to_yaml() +
+               " height: " + std::to_string(height) + " width: " + std::to_string(width) + "}";
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const AnchoredRectangle& rc);
+};
+
+std::ostream& operator<<(std::ostream& os, const AnchoredRectangle& rc) {
+    os << rc.to_string();
+    return os;
+}
+
+// Given an index component `idx_val`, and a radius `r` around it, returns the
+// start and end index components, and length of the range, that fit in the
+// [0, max_range] limits.
+inline std::tuple<int, int, int> centered_range(int idx_val, const int r, const int max_range) {
+    // pin start to the []0, max_range] range
+    int start = std::max(0, idx_val - r);
+    start = std::min(start, max_range);
+
+    // pin end to the [0, max_range]
+    int end = std::max(0, idx_val + r);
+    end = std::min(max_range, idx_val + r);
+
+    // range is inclusive of the first element, and can not be longer than start
+    // minus max range
+    int length = end - start + 1;
+    length = std::min(length, max_range - start);
+
+    return std::make_tuple(start, end, length);
+}
+
+// get an Eigen block coordinates (top-left, height, width) anchored inside a
+// square matrix of dimensions 2*r+1 (anchor top-left).
+inline AnchoredRectangle anchored_block(const Index& idx, const int r, const unsigned height,
+                                        const unsigned width) {
+    auto [top, bot, rangey] = centered_range(idx.j, r, height);
+    auto [left, right, rangex] = centered_range(idx.i, r, width);
+    assertm(rangey > 0, "Selected block lies outside of the image limits.");
+    assertm(rangex > 0, "Selected block lies outside of the image limits.");
+
+    int anchor_top = std::max(r - idx.j, 0);
+    int anchor_left = std::max(r - idx.i, 0);
+
+    // now it's safe to cast ranges to unsigned
+    return {{top, left}, {anchor_top, anchor_left}, (unsigned)rangey, (unsigned)rangex};
+}
+
+inline auto manhattan_neighbors(const Index& idx, const unsigned width, const unsigned height) {
```

Comment:
Should this be height then width for consistency?

Reply:
I guess, somewhat annoyingly width and height would depend on the method signature - i.e. is it an Index or a Point, no?

I kept all references (and reverted changes) to `(width, height)` because that's the general graphics field convention and we call our objects Images a lot and I don't want that bit of ordering be dependent on whether we're taking an `Index` or an `Point`. I flip the `np.array.shape` ordering in the bindings for the methods that take a tuple, but short of either making everything consistently always `y, x` because we're dealing with arrays, I don't see an easier solution to this.